File: apps/frontend/src/entities/art/shipment.ts
import { Availability, Dimensions, ISODate, Localized, Money } from '@/entities/common';
import { ImagesJSON } from './images';

export type ImageShipment =
    | {
          kind: 'hopper';
          hopperSrc: string; // file url in the hopper
      }
    | {
          kind: 'ready';
          image: ImagesJSON;
      };

export interface ArtShipment {
    readonly id: string;
    dateCreated: ISODate;
    title?: Localized;
    techniques: string[];
    availability: Availability;
    dimensions: Dimensions;
    price?: Money;
    alt?: Localized;
    series?: string;
    tags: string[];
    notes?: string;
    images: ImageShipment;
}

File: apps/admin-backend/app/models/shipment.py
from typing import Literal, Union, Annotated, Optional, List
from pydantic import BaseModel, Field

from app.models.images_pipline import ImagesJSON
from .common import Availability, Dimensions, ISODate, Localized, Money


class HopperImageShipment(BaseModel):
    kind: Literal["hopper"]
    hopperSrc: str


class ReadyImageShipment(BaseModel):
    kind: Literal["ready"]
    image: ImagesJSON


ImageShipment = Annotated[
    Union[HopperImageShipment, ReadyImageShipment], Field(discriminator="kind")
]


class ArtShipmentModel(BaseModel):
    id: str
    dateCreated: ISODate
    title: Optional[Localized] = None
    techniques: Optional[List[str]] = None
    availability: Optional[Availability] = None
    dimensions: Optional[Dimensions] = None
    price: Optional[Money] = None
    alt: Optional[Localized] = None
    series: Optional[str] = None
    tags: Optional[List[str]] = None
    notes: Optional[str] = None
    images: ImageShipment
